### ISSUE TITLE
Group operations are not checking if student is already in a group

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#8394]
To reproduce:
* open a group assessment;
* change effective user to a student;
* duplicate the tab;
* in one tab create a group or join an existing group;
* in the other tab, create a group or join another existing group.

The second operation succeeds in creating/joining the group (it shouldn't). After this point, the student is unable to open the assessment, since the queries for assessment assume a single group exists for the student.

So, in essence, the issue is that creating a group and joining a group are not currently validated to ensure a student is only part of one group. A typical workflow does not allow this to happen, since a user will not see the option to join/create a group if they are already part of it, but using multiple tabs/browsers or abusing the history (back button) can cause this to be a viable option.

A related issue (reported by @glherman) seems to be a case where joining a group causes an error but ends up adding the student to the group anyway. Due to lack of additional information it's hard to pinpoint the exact issue, but fixing the bug above should keep this bug from causing unintended consequences anyway.

**Comment by nwalters512** [comment]
We should really be enforcing this with `UNIQUE` constraints. Unfortunately, the table that tracks group membership (`group_users`) doesn't have any context about which assessment a given group membership belongs to. We might consider adding a `group_config_id` column to that table and enforcing uniqueness on `(user_id, group_config_id)`.

We can of course do this without database constraints; this is just the perfect kind of invariant to enforce automatically.

**Comment by nwalters512** [comment]
The following query will show all the `group_users` rows that currently violate our desired uniqueness constraint:

```sql
SELECT
  gc.id AS group_configs_id,
  gc.course_instance_id AS course_instance_id,
  gc.assessment_id AS assessment_id,
  gu.user_id AS user_id,
  u.uid AS user_uid,
  ARRAY_AGG(g.name) AS group_names,
  ARRAY_AGG(g.id) AS group_ids,
  COUNT(gu.*)
FROM
  group_users AS gu
  JOIN groups AS g ON (g.id = gu.group_id)
  JOIN group_configs AS gc ON (gc.id = g.group_config_id)
  JOIN users AS u ON (u.user_id = gu.user_id)
GROUP BY
  gc.id,
  gc.course_instance_id,
  gc.assessment_id,
  gu.user_id,
  u.uid
HAVING
  COUNT(gu.*) > 1;
```

This yields 335 rows in `us-prod` and 143 rows in `ca`, where each row is a user with more than one group membership for a given assessment. We'll have to clean these up before we can enforce uniqueness in the database itself.

Modifying that query, we can identify that many of these are trival to fix:

```sql
SELECT
  gc.id AS group_configs_id,
  gc.course_instance_id AS course_instance_id,
  gc.assessment_id AS assessment_id,
  gu.user_id AS user_id,
  u.uid AS user_uid,
  ARRAY_AGG(g.name) AS group_names,
  ARRAY_AGG(g.id) AS group_ids,
  ARRAY_AGG(ai.id) AS assessment_instance_ids,
  COUNT(gu.*) AS num_memberships,
  num_nonnulls (VARIADIC ARRAY_AGG(ai.id)) AS num_assessment_instances
FROM
  group_users AS gu
  JOIN groups AS g ON (g.id = gu.group_id)
  JOIN group_configs AS gc ON (gc.id = g.group_config_id)
  JOIN users AS u ON (u.user_id = gu.user_id)
  LEFT JOIN assessment_instances AS ai ON (ai.group_id = g.id)
GROUP BY
  gc.id,
  gc.course_instance_id,
  gc.assessment_id,
  gu.user_id,
  u.uid
HAVING
  COUNT(gu.*) > 1
ORDER BY
  num_assessment_instances DESC;
```

For any user/assessment pair where `num_assessment_instances = 1`, we can trivially delete the group membership for all groups without an assessment instances. This is the vast majority of duplicate memberships: 292 rows in `us-prod`, and 127 rows in `ca`.

This leaves us with some remaining users who either have zero assessment instances or more than one assessment instances. I'm not quite sure what the correct thing to do with these is. Spot-checking a few cases, it seems like a common one is students end up with a group by themselves (with low credit) and a group with other people (high credit). We could delete on those criteria: keep the group with the highest score, or if scores are the same, keep the group where the user is with other people (as opposed to by themselves). I can write some more queries tomorrow to see how close those heuristics would get us to fully eliminating multiple group memberships.

**Comment by nwalters512** [comment]
Here's an enhanced query that shows assessment scores and group sizes:

```sql
SELECT
  gc.id AS group_configs_id,
  gc.course_instance_id AS course_instance_id,
  gc.assessment_id AS assessment_id,
  gu.user_id AS user_id,
  u.uid AS user_uid,
  ARRAY_AGG(g.name) AS group_names,
  ARRAY_AGG(g.id) AS group_ids,
  ARRAY_AGG(ai.id) AS assessment_instance_ids,
  ARRAY_AGG(ai.score_perc) AS assessment_scores,
  ARRAY_AGG(
    (
      SELECT
        COUNT(*)
      FROM
        group_users
      WHERE
        group_id = g.id
    )
  ) AS group_sizes,
  COUNT(gu.*) AS num_memberships,
  num_nonnulls (VARIADIC ARRAY_AGG(ai.id)) AS num_assessment_instances
FROM
  group_users AS gu
  JOIN groups AS g ON (g.id = gu.group_id)
  JOIN group_configs AS gc ON (gc.id = g.group_config_id)
  JOIN users AS u ON (u.user_id = gu.user_id)
  LEFT JOIN assessment_instances AS ai ON (ai.group_id = g.id)
GROUP BY
  gc.id,
  gc.course_instance_id,
  gc.assessment_id,
  gu.user_id,
  u.uid
HAVING
  COUNT(gu.*) > 1
ORDER BY
  num_assessment_instances DESC;
```

Based on the data I saw with this, I propose the following strategy to remove duplicate memberships:

- First pass: find all duplicate group memberships that don't have any corresponding assessment instances, order by group size, and remove memberships but that to the largest group. If multiple groups have the same size, retain only the one with the larger `assessment_instance.id`.
- Second pass: for all duplicate group memberships, remove all memberships to groups that don't have a corresponding assessment instance.
- Third pass: for all duplicate group memberships, order the assessment instances by score and remove memberships to all groups but the one with the highest-scoring instance. If multiple instances have the same score, retain the membership in the group with the larger size. If multiple groups have the same size, retain the one with the larger `assessment_instance.id`.

**Comment by jonatanschroeder** [comment]
> * First pass: find all duplicate group memberships that don't have any corresponding assessment instances, order by group size, and remove memberships but that to the largest group. If multiple groups have the same size, retain only the one with the larger `assessment_instance.id`.

Do you mean `group.id`? If there is no assessment instance, there is no assessment instance ID.

**Comment by nwalters512** [comment]
@jonatanschroeder yes, good catch - I updated my comment.

**Comment by jonatanschroeder** [comment]
I assume this should be done *after* the group joining code starts checking for existing group membership?

**Comment by nwalters512** [comment]
@jonatanschroeder yeah, we both had the same idea at the same time, I edited my comment right before you posted that.

**Comment by nwalters512** [comment]
I implemented manual uniqueness checking here: #8905
